### PR TITLE
rfc: kv lock observability via `crdb_locks`

### PR DESCRIPTION
**Pull request by AlexTalks**
_Link to [RFC text](https://github.com/AlexTalks/cockroach/blob/rfc_crdb_locks/docs/RFCS/20220104_crdb_locks.md)._

This RFC proposes a feature for observing locks (and contention on
locks) via a new virtual table called `crdb_internal.crdb_locks`.

Release note: None

### CODE REVIEW

**Comment by AlexTalks** [comment]
Formatted view: https://github.com/AlexTalks/cockroach/blob/rfc_crdb_locks/docs/RFCS/20220104_crdb_locks.md

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/RFCS/20220104_crdb_locks.md, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuNIVu9DRiG5aAO8Lps:-MuNIVu9DRiG5aAO8Lpt:b-xw9c73) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0b408cfd02590026d91825dbd58c2a4b6bc94d9/docs/RFCS/20220104_crdb_locks.md#L137)):*
> ```Markdown
> short-lived and only held as long as a single request, rather than for the life 
> of a transaction, and thus will be less useful in visualizing contention.
> 4. **We do not track the queries or statements that obtain locks.**  This is a 
> ```

Tagging @cockroachdb/sql-observability as I want to make sure this assumption is valid given current capabilities.  IIUC, we don't expose the statements in a given transaction currently, but given that the `crdb_locks` table is being added, perhaps this is something that could be useful going forward, in order to visualize the statement that caused a transaction to obtain a given lock?

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @maryliag)

___
*[docs/RFCS/20220104_crdb_locks.md, line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuNIVu9DRiG5aAO8Lps:-MuRAN7dBxtfWWLUfgfP:b-k6pxyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0b408cfd02590026d91825dbd58c2a4b6bc94d9/docs/RFCS/20220104_crdb_locks.md#L137)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Tagging @cockroachdb/sql-observability as I want to make sure this assumption is valid given current capabilities.  IIUC, we don't expose the statements in a given transaction currently, but given that the `crdb_locks` table is being added, perhaps this is something that could be useful going forward, in order to visualize the statement that caused a transaction to obtain a given lock?
</blockquote></details>

Your assumption is correct. Currently we don't track which statement caused a lock.

**Comment by erikgrinaker** [comment]
Great to see this!

Just a drive-by comment: I've been wanting a tree view of contention, since we often have one long-running txn that's blocking a bunch of others, and those in turn are blocking others again, etc. Some way to easily visualize the entire dependency chain and identify the root transaction causing the pileup seems like it'd be really useful.

I think that should be doable with this proposal using a recursive CTE? The only issues I see is that there could be holes in the chain -- both due to the fact that we'll have an inconsistent view of the contention (since we're independently querying the in-memory state of several nodes at different times), and possibly because we perhaps aren't tracking the entire set of involved transactions in the lock table. Thoughts on this?

**Review by Azhng — COMMENTED** [review]
Nice! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-:-Muw-NSFBcybeWRgPnPr:bwuv9up)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, and @maryliag)

___
*[docs/RFCS/20220104_crdb_locks.md, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvtJPv92lEwKm7Uywt:-MuvtJPv92lEwKm7Uywu:b-mw5wak) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L50)):*
> ```Markdown
> | Term                     | Explanation |
> | ------------------------ | ----------- |
> | Sequencing               | Processing a request through Currency Control.       |
> ```

nit: s/Currency/Concurrency/
___
*[docs/RFCS/20220104_crdb_locks.md, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvuIlz9DmU9lOsbIhk:-MuvuIlz9DmU9lOsbIhl:b-xhpbgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L58)):*
> ```Markdown
> | Replicated Lock          | A durable lock represented in persistent storage by what is commonly known as an “intent” (`TxnMeta`).  This type of lock may or may not be tracked in the in-memory lock table - as mentioned above, a replicated lock is tracked in the lock table if there are wait queues made up of other transactions waiting to acquire this lock.         |
> | Unreplicated Lock        | A lock that is only represented in the lock table, and will not persist in case of node failure or lease transfer.  Used in `[Get/Scan]ForUpdate`.        |
> | Replicated Lock Keyspace | The separate keyspace in which replicated locks, commonly known as “separated intents”, are stored; in the Local keyspace of a range.  This is entirely separate from the Lock Table.  Example intent key in the replicated lock keyspace: `/Local/Lock/Intent/Table/56/1/1169/5/3054/0`.        |
> ```

Is it correct to understand that even thought it's called a "Local" keyspace, the lock is still replicated ? 
___
*[docs/RFCS/20220104_crdb_locks.md, line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvvtYLBc6cIThfygUx:-MuvvtYLBc6cIThfygUy:b-c31aiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L141)):*
> since we do not keep track of the statement history in a given transaction

Are we talking about the current active transaction? Or historical transaction? If it's latter, then we do track list of statements in that transaction.

For former, currently there's no interface to query that interface, but we do have that information stored somewhere (e.g. in `connExecutor` we keep track of list of statement fingerprint ID of an active txn so far) and it can be possibly exposed via a SQL interface. 
___
*[docs/RFCS/20220104_crdb_locks.md, line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Muw-5iA4BYXuNvivghE:-Muw-5iA4BYXuNvivghF:b50dzzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L237)):*
> ```Markdown
> 
> 	repeated LockWaiter lock_waiters = 7;
> }
> ```

small style nits: seems like tabs and spaces are mixed up in here

**Comment by Azhng** [comment]
Another quick question just for my own understanding. When the lease moves from one node to another, what happens to the in-memory `lockTable`? Does it get serialized and sent to the new leaseholder?

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, and @maryliag)

___
*[docs/RFCS/20220104_crdb_locks.md, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvtJPv92lEwKm7Uywt:-Mux6pGv2EU9V8NqppiX:b-4k90tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L50)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: s/Currency/Concurrency/
</blockquote></details>

HA, good catch!
___
*[docs/RFCS/20220104_crdb_locks.md, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvuIlz9DmU9lOsbIhk:-Mux77Ta2FsvYRgEwfyJ:b6he6zy) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L58)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Is it correct to understand that even thought it's called a "Local" keyspace, the lock is still replicated ? 
</blockquote></details>

That is correct.  "Local" in this context means that this key is only addressable within the context of this (replicated) range.
___
*[docs/RFCS/20220104_crdb_locks.md, line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-MuvvtYLBc6cIThfygUx:-Mux7fXB3nKAVovNuSik:b-co9wie) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L141)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

> since we do not keep track of the statement history in a given transaction

Are we talking about the current active transaction? Or historical transaction? If it's latter, then we do track list of statements in that transaction.

For former, currently there's no interface to query that interface, but we do have that information stored somewhere (e.g. in `connExecutor` we keep track of list of statement fingerprint ID of an active txn so far) and it can be possibly exposed via a SQL interface. 
</blockquote></details>

I was speaking about the statement history of the current active transaction - can clarify that.

As for the ability to possibly expose it, we should schedule some time soon to discuss this as it would be a great enhancement.
___
*[docs/RFCS/20220104_crdb_locks.md, line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Muw-5iA4BYXuNvivghE:-Mux8BKp3yL8N3oKw5hU:bd5l6mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L237)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

small style nits: seems like tabs and spaces are mixed up in here 
</blockquote></details>

thanks!

**Comment by AlexTalks** [comment]
> Another quick question just for my own understanding. When the lease moves from one node to another, what happens to the in-memory `lockTable`? Does it get serialized and sent to the new leaseholder?

**tl;dr** it does not get serialized and sent to the leaseholder, rather things eventually get redirected and queues re-form.

Longer explanation: It does not get serialized and sent to the new leaseholder - rather, in the case of a lease transfer, all locks are cleared from the in-memory lock table and the lock table is disabled.  On clearing the locks, the waiters are (in order) notified that they no longer need to wait and can continue - however, as they attempt to proceed to command evaluation, any conflicting reads/writes will end up encountering the replicated intent, thus resulting in a `WriteIntentError` which means the request needs to be re-sequenced through Concurrency Control.  After re-sequencing the request (which will trivially pass without waiting, since the lock table is disabled), before attempting to evaluate [the lease is re-checked](https://github.com/cockroachdb/cockroach/blob/6664d0c34df0fea61de4fff1e97987b7de609b9e/pkg/kv/kvserver/replica.go#L1327), and upon discovering that this replica is no longer the leaseholder, the request should thus be redirected to the proper leaseholder replica for the range.  When this happens, the transaction will re-enter the wait queue if necessary.

One illustrative example exists in the [range_state_listener test](https://github.com/cockroachdb/cockroach/blob/55720d0a4c0dd8030cd19645461226d173eefc2e/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/range_state_listener).

**Comment by AlexTalks** [comment]
In response to @erikgrinaker:
> I think that should be doable with this proposal using a recursive CTE? The only issues I see is that there could be holes in the chain -- both due to the fact that we'll have an inconsistent view of the contention (since we're independently querying the in-memory state of several nodes at different times), and possibly because we perhaps aren't tracking the entire set of involved transactions in the lock table. Thoughts on this?

I think that the issue about the potential for inconsistency is unavoidable, as I don't think we'd want to block anything to render this table.  As for tracking the entire set of involved transactions, I think any that any waiters on a blocking transaction should have that blocking transaction as the lock holder for that key - so if txn1 holds key A but waits on key B, and txn2 holds key B but waits on key C (held by txn3), we should be able to visualize the full dependency chain through the lockholders `txn1 -> txn2 -> txn3` (if this sounds incorrect, please let me know!)

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @maryliag)

___
*[docs/RFCS/20220104_crdb_locks.md, line 237 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Muw-5iA4BYXuNvivghE:-Mv-qk0i6Uba_ZqZUSQc:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/452b0ebe687c344ed970ca16c9b1d7cce4a2948f/docs/RFCS/20220104_crdb_locks.md#L237)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

thanks!
</blockquote></details>

OK

**Review by nvb — APPROVED** [review]
:lgtm: nice job on this Alex!

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-:-Mv-ysDM5IMaOe_H5f9c:b-pocn79)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks, @Azhng, @maryliag, and @nvanbenschoten)

___
*[docs/RFCS/20220104_crdb_locks.md, line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-vl5iENOKkMccK4nH:-Mv-vl5iENOKkMccK4nI:b-kztgl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L123)):*
> ```Markdown
> be acceptable given the use case of this feature.
> 
> 1. **Inability to display replicated locks without waiters**.  Replicated locks 
> ```

This is framed as a limitation, but that's quite right. This is an intentional design decision. We could scan the persistent lock table and include locks all locks in the response to a `QueryLocksRequest` if we wanted to. In fact, we very well may want to include that as an option on the request. But for the same reason as in PG, we've deemed the IO cost of doing so to be too expensive for the `crdb_internal.crdb_locks` table.
___
*[docs/RFCS/20220104_crdb_locks.md, line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-vnoR5gNXA6JuRAwN:-Mv-vnoR5gNXA6JuRAwO:bvfkbdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L128)):*
> ```Markdown
> storage engine.  Given that this feature is intended to visualize contention, 
> this may be of minimal concern.
> 2. **Only read-write transactional lockholders can be displayed.**  While 
> ```

Isn't this also a design decision? We could return the requests in a `lockWaitQueue.waitingReaders` set if we wanted to. We decided not to do this for pg_locks because it would be confusing to see non-locking operations in that table, given its contract and based on how concurrency control works in Postgres. But are we sure we want to make that decision for `crdb_locks`, where we have more flexibility to write the script? If read-only requests still block on locks, should they be represented somehow in `crdb_locks`?
___
*[docs/RFCS/20220104_crdb_locks.md, line 170 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-wmZtBO4OTJ1jJngb:-Mv-wmZtBO4OTJ1jJngc:bfhkwk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L170)):*
> ```Markdown
> CREATE TABLE crdb_internal.crdb_locks (
>   range_id INT,           -- the ID of the range that contains the lock
>   table_id INT,           -- the id of the table to which the range with this lock belongs
> ```

Should we include `database_id` and `index_id` in this table?
___
*[docs/RFCS/20220104_crdb_locks.md, line 210 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-xUjfEr3w_IIMnVsb:-Mv-xUjfEr3w_IIMnVsc:b-o222zy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L210)):*
> ```Markdown
> 
> 
> ```
> ```

nit: "```proto"
___
*[docs/RFCS/20220104_crdb_locks.md, line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-xi-zADmI8OHY8PZf:-Mv-xi-zADmI8OHY8PZg:b-aehnbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L220)):*
> ```Markdown
>   bool active_waiter = 2;
>   SpanAccess access = 3; // [ReadOnly, ReadWrite]
>   uint64 seq_num = 4;
> ```

I don't think we should include `seq_num` in this struct. It's an implementation detail of the lockTable data structure that does not make sense to expose.
___
*[docs/RFCS/20220104_crdb_locks.md, line 232 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-xzNY9hDdZWahEhIA:-Mv-xzNY9hDdZWahEhIB:bz0ga7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L232)):*
> ```Markdown
>   SpanScope span_scope = 2; // [GLOBAL, LOCAL]
>   Strength lock_strength = 3;
>   Durability = 4;
> ```

`Durability durability = 4;`
___
*[docs/RFCS/20220104_crdb_locks.md, line 236 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-xtr-60F_DT3oEMfP:-Mv-xtr-60F_DT3oEMfQ:b2kwhqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L236)):*
> ```Markdown
>   TxnMeta lockholder_txn_meta = 6;
> 
>   repeated LockWaiter lock_waiters = 7;
> ```

Is this list in a specific order?
___
*[docs/RFCS/20220104_crdb_locks.md, line 435 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-ycWXD_m9Yhxog_rD:-Mv-ycWXD_m9Yhxog_rE:b-5fgkxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L435)):*
> ```Markdown
> | **For Users/DB Admins** | Contention Events (via SQL, Dashboards) | `crdb_locks` (via SQL) |
> 
> ## Examples
> ```

These are cool to see written out!
___
*[docs/RFCS/20220104_crdb_locks.md, line 518 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-ynn02zN_Xg_UhPeJ:-Mv-ynn02zN_Xg_UhPeK:b2dzg2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L518)):*
> ```Markdown
> # Open Questions
> 
> * Special considerations for tenant SQL pods in Serverless
> ```

What is the question here? Do we expect things to behave differently in serverless?

**Comment by erikgrinaker** [comment]
> I think that the issue about the potential for inconsistency is unavoidable, as I don't think we'd want to block anything to render this table.

I think that's true. We may want to mention this as a limitation in the RFC and any documentation, but I don't think it's likely to reduce its usefulness -- more of a potential gotcha.

> As for tracking the entire set of involved transactions, I think any that any waiters on a blocking transaction should have that blocking transaction as the lock holder for that key - so if txn1 holds key A but waits on key B, and txn2 holds key B but waits on key C (held by txn3), we should be able to visualize the full dependency chain through the lockholders `txn1 -> txn2 -> txn3` (if this sounds incorrect, please let me know!)

I seem to recall something about the lock table having a bounded size and evicting certain locks when it's full, but I don't remember the details. It's possible (probable even) that we'll pull contended locks into memory even if the lock table is otherwise full. Just worth pointing it out if there is a limitation here.

Thanks again working on this -- RFC looks great, and I'm thrilled to see progress on this.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75541)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, and @nvanbenschoten)

___
*[docs/RFCS/20220104_crdb_locks.md, line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-vl5iENOKkMccK4nH:-Mv0NxCK9jSlifdqRkCz:btuz815) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L123)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is framed as a limitation, but that's quite right. This is an intentional design decision. We could scan the persistent lock table and include locks all locks in the response to a `QueryLocksRequest` if we wanted to. In fact, we very well may want to include that as an option on the request. But for the same reason as in PG, we've deemed the IO cost of doing so to be too expensive for the `crdb_internal.crdb_locks` table.
</blockquote></details>

Agreed.  I think some of the limitations may also be seen as a clarification of semantics.  Let me know if you think anything is needed to add here though!
___
*[docs/RFCS/20220104_crdb_locks.md, line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-vnoR5gNXA6JuRAwN:-Mv0OaaU6VPBGGGm7CZj:bc402ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L128)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Isn't this also a design decision? We could return the requests in a `lockWaitQueue.waitingReaders` set if we wanted to. We decided not to do this for pg_locks because it would be confusing to see non-locking operations in that table, given its contract and based on how concurrency control works in Postgres. But are we sure we want to make that decision for `crdb_locks`, where we have more flexibility to write the script? If read-only requests still block on locks, should they be represented somehow in `crdb_locks`?
</blockquote></details>

This limitation was only about lockholders, not waiters - I see no reason not to include read-only lock waiters in the response of course!
___
*[docs/RFCS/20220104_crdb_locks.md, line 170 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-wmZtBO4OTJ1jJngb:-Mv0OtO_-r85BK30tpnC:bp3c7vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L170)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we include `database_id` and `index_id` in this table?
</blockquote></details>

Sure, can add these!
___
*[docs/RFCS/20220104_crdb_locks.md, line 236 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-xtr-60F_DT3oEMfP:-Mv0PTZY-rwfL2mTt0Km:b-dn3c7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L236)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this list in a specific order?
</blockquote></details>

You mean in order of readers vs writers?  I think it would be ordered: `(reservation, waiting readers, waiting writers)`, but if it would be better to separate these out that could be done.
___
*[docs/RFCS/20220104_crdb_locks.md, line 435 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-ycWXD_m9Yhxog_rD:-Mv0PiI_1CJpeituI7k2:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L435)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These are cool to see written out!
</blockquote></details>

Thanks!
___
*[docs/RFCS/20220104_crdb_locks.md, line 518 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75541#-Mv-ynn02zN_Xg_UhPeJ:-Mv0PkJHDZ4Fk95YrzCb:b-4s9nct) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb0408641551e9e49e1cfe94a6846abbd5571f1/docs/RFCS/20220104_crdb_locks.md#L518)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What is the question here? Do we expect things to behave differently in serverless?
</blockquote></details>

The main consideration is how to account for the request in the billing model I think.  Or, essentially, if there are any "unknown" unknowns w/r/t Serverless that aren't otherwise accounted for in this document.

**Comment by nvanbenschoten** [comment]
@jordanlewis you mentioned today that there is limited support for predicate pushdown in virtual tables. Could you point us at that? The `crdb_internal.crdb_locks` proposed here would meaningfully benefit from predicate pushdown on the `table_id` column.

**Comment by jordanlewis** [comment]
Take a look at how other virtual tables use the `indexes` field of the `virtualSchemaTable` struct. Each virtual index is implemented with a `populate` method that takes a `Datum` that represents the constrained column. They only support a single column equality constraint at the moment.

**Comment by cjireland** [comment]
I may have missed this but will this information be available after the event? It's OK whilst a TXN is running but typically you need to see this lock information as part of an investigation (for example: we are seeing a lot of contention at 10am each day). In a complex, event driven, system the interactions may be difficult to take apart and even reproduce repeatedly. Recourse to this table after, say a test where contention was experienced, would be very helpful.
